### PR TITLE
Add platform view support for linux

### PR DESCRIPTION
**Pull request by huanghongxun**
This pull request is part of #24169. For framework API, see https://github.com/flutter/flutter/pull/74814.

This pull request adds platform view API to Linux shell, without support for touch events and clipping mutations, which are splitted and proposed later when this PR get landed.

Code changes:
1. `FlPlatformView` to wrap a GtkWidget, owned by `FlPlatformViewsPlugin`. This abstraction allows us to owning a `FlMethodChannel` in `FlPlatformView` for communication with dart codes.
2. `FlPlatformViewFactory` for creating platform views for specific view type. Flutter Framework will call functions of `FlPlatformViewFactory` when creating a new platform view.
3. `FlPlatformViewsPlugin` for implementing `SystemChannels.platform_views` from Flutter services library.

A working demo with webview: https://github.com/wechat-miniprogram/gtktest.

This pr does not support applying mutations on the platform views yet. I will open a PR later to implement that feature.

This pr is part of platform view implementation on Linux, related issue: flutter/flutter#41724.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by huanghongxun** [comment]
@robert-ancell @iskakaushik This PR contains parts other than `FlGestureHelper` and `FlClippingView` from #24169. Changes on `FlView`  are simplified.

**Comment by chinmaygarde** [comment]
cc @gw280 as well. I know you are looking at the Mac side of things but it may be a good idea to go over this once to look for common patterns.

**Comment by wanjm** [comment]
when will it be merged?

**Comment by wanjm** [comment]
why does it blocked here

**Comment by huanghongxun** [comment]
@robert-ancell Please take a look at this pr.

File: shell/platform/linux/fl_platform_views_plugin_test.cc [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                                  cb, loop);
+}
+
+static void create_platform_view(GMainLoop* loop,
+                                 FlBinaryMessenger* messenger,
+                                 int64_t view_id,
+                                 const gchar* view_type,
+                                 FlValue* args,
+                                 GAsyncReadyCallback cb) {
+  FlValue* create_args = fl_value_new_map();
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin_test.cc:58]: Missing g_autoptr

File: shell/platform/linux/fl_view_private.h [hunk 2]
```diff
@@ -8,6 +8,17 @@
 #include "flutter/shell/platform/linux/public/flutter_linux/fl_view.h"
 
 #include "flutter/shell/platform/linux/fl_gl_area.h"
+#include "flutter/shell/platform/linux/fl_platform_views_plugin.h"
+
+/**
+ * fl_view_get_platform_views_plugin(FlView* view);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_view_private.h:14]: Should be just `fl_view_get_platform_views_plugin:`

**Review by robert-ancell — APPROVED** [review]
Two small fixes required, but otherwise looks great. Thanks @huanghongxun!

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    send_response(method_call, response);
+  }
+}
+
+static void fl_platform_views_plugin_dispose(GObject* object) {
+  FlPlatformViewsPlugin* self = FL_PLATFORM_VIEWS_PLUGIN(object);
+
+  g_clear_object(&self->channel);
+  g_hash_table_destroy(self->factories);
+  g_hash_table_destroy(self->platform_views);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:258]: Can you change these two lines to:
```c
g_clear_pointer(&self->factories, g_hash_table_unref);
g_clear_pointer(&self->platform_views, g_hash_table_unref);
```

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  g_autoptr(FlStandardMethodCodec) codec = fl_standard_method_codec_new();
+  self->channel =
+      fl_method_channel_new(messenger, kChannelName, FL_METHOD_CODEC(codec));
+  fl_method_channel_set_method_call_handler(self->channel, method_call_cb, self,
+                                            nullptr);
+
+  self->factories =
+      g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_object_unref);
+  self->platform_views = g_hash_table_new_full(g_direct_hash, g_direct_equal,
+                                               nullptr, g_object_unref);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:287]: Please move into `fl_platform_views_plugin_init` for consistency with other classes.

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+  return self;
+}
+
+gboolean fl_platform_views_plugin_register_view_factory(
+    FlPlatformViewsPlugin* self,
+    FlPlatformViewFactory* factory,
+    const gchar* view_type) {
+  g_object_ref(factory);
+  return g_hash_table_insert(self->factories, g_strdup(view_type), factory);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:297]: Reads a little clearer as:
```c
return g_hash_table_insert(self->factories, g_strdup(view_type), g_hash_table_ref(factory));
```

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+static constexpr char kSetDirectionMethod[] = "setDirection";
+static constexpr char kEnterMethod[] = "enter";
+static constexpr char kExitMethod[] = "exit";
+
+struct _FlPlatformViewsPlugin {
+  GObject parent_instance;
+
+  FlMethodChannel* channel;
+
+  GHashTable* factories;  // string -> FlPlatformViewFactory*
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:29]: Can you please add a comment above this about what this value is.

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+  GHashTable* platform_views;  // int -> FlPlatformView*
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:31]: Can you please add a comment above this about what this value is.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Changing status as a few small things to fix.

File: shell/platform/linux/fl_platform_views_plugin_test.cc [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+  g_autoptr(FlEngine) engine = make_mock_engine();
+  g_autoptr(FlBinaryMessenger) messenger = fl_binary_messenger_new(engine);
+
+  fl_binary_messenger_set_message_handler_on_channel(
+      messenger, "test/responses", response_cb, loop, nullptr);
+
+  g_autoptr(FlPlatformViewsPlugin) plugin =
+      fl_platform_views_plugin_new(messenger);
+  FlMockViewFactory* factory =
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin_test.cc:125]: Change to `g_autoptr`, see other comment about this.

File: shell/platform/linux/fl_plugin_registrar.cc [hunk 9]
```diff
@@ -71,3 +72,19 @@ G_MODULE_EXPORT FlView* fl_plugin_registrar_get_view(FlPluginRegistrar* self) {
...
+    const gchar* view_type) {
+  g_return_if_fail(FL_IS_PLUGIN_REGISTRAR(self));
+
+  if (self->view) {
+    FlPlatformViewsPlugin* plugin =
+        fl_view_get_platform_views_plugin(self->view);
+    fl_platform_views_plugin_register_view_factory(plugin, factory, view_type);
+  } else {
+    g_warning("Cannot register view factory before FlView is initialized");
+    g_object_unref(factory);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_plugin_registrar.cc:88]: Remove this.

File: shell/platform/linux/public/flutter_linux/fl_platform_views.h [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+/**
+ * fl_platform_view_factory_create_platform_view:
+ * @factory: an #FlPlatformViewFactory.
+ * @view_identifier: a unique identifier for this #GtkWidget.
+ * @args: parameters for creating the #GtkWidget sent from the Dart side of
+ *        the Flutter app. If get_create_arguments_codec is not implemented, or
+ *        if no creation arguments were sent from the Dart code, this will be
+ *        null. Otherwise this will be the value sent from the Dart code as
+ *        decoded by get_create_arguments_codec.
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_platform_views.h:197]: These last two lines don't make sense here (copy/paste from above). They can be removed.

File: shell/platform/linux/public/flutter_linux/fl_platform_views.h [hunk 11]
```diff
@@ -0,0 +1,224 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_platform_views.h:1]: Can you split this file into `fl_platform_view.h` and `fl_platform_view_factory.h`? It's easier to understand with one class per header.

File: shell/platform/linux/fl_platform_views.cc [hunk 12]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views.cc:1]: Please split into `fl_platform_view.cc` and `fl_platform_view_factory.cc`

File: shell/platform/linux/fl_platform_views_plugin.cc [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+      FL_PLATFORM_VIEW_FACTORY(g_hash_table_lookup(self->factories, view_type));
+  if (!factory) {
+    return FL_METHOD_RESPONSE(fl_method_error_response_new(
+        kBadArgumentsError, "View type does not exist", nullptr));
+  }
+
+  g_autoptr(FlMessageCodec) codec =
+      fl_platform_view_factory_get_create_arguments_codec(factory);
+  FlValue* creation_params = nullptr;
+  if (codec && params_value) {
```
Comment (robert-ancell) [inline shell/platform/linux/fl_platform_views_plugin.cc:96]: Should we report an error or warning if `params_value != NULL` and `code == NULL`? i.e. it currently would silently lose the parameters. I'm really wondering if it's worth the effort having the codec requested here - why not pass the parameters as a GBytes to `fl_platform_view_factory_create_platform_view` and let the factory work out how to decode them? What do the other Flutter implementations do?

File: shell/platform/linux/fl_platform_views_plugin_test.cc (hunk 1 above)
Comment (huanghongxun) [reply shell/platform/linux/fl_platform_views_plugin_test.cc:58]: `invoke_method` takes ownership of `create_args` here. So we cannot free here.

File: shell/platform/linux/public/flutter_linux/fl_platform_views.h (hunk 10 above)
Comment (huanghongxun) [reply shell/platform/linux/public/flutter_linux/fl_platform_views.h:197]: I think that we can hide `fl_platform_view_factory_create_platform_view` from plugins, and can be moved to `fl_platform_views_private.h`.

File: shell/platform/linux/fl_platform_views_plugin.cc (hunk 13 above)
Comment (huanghongxun) [reply shell/platform/linux/fl_platform_views_plugin.cc:96]: iOS implementation does not print any logs when `args[@"params"]` (`params_value` in Linux here) is null.
https://github.com/flutter/engine/blob/2550df40bd9a8825e9b7b46d071c35faf2748f4e/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm#L149-L156

Android implementation does the same.
https://github.com/flutter/engine/blob/2550df40bd9a8825e9b7b46d071c35faf2748f4e/shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java#L146-L149

**Comment by wanjm** [comment]
anything blocked?

**Comment by huanghongxun** [comment]
> anything blocked?

It seems that few people can review this and the dart side pull requests. And these people have no time.

**Comment by dkwingsmt** [comment]
Maybe @cbracken and @gspencergoog ?

**Comment by cbracken** [comment]
@huanghongxun it's been a long while, but is this patch something you're willing to get rebased and tests fixed or would you like someone to pick this up and continue with it? Unfortunately we've been pretty short on people who could spend the appropriate amount of review time on this (and https://github.com/flutter/flutter/pull/74814) but I've got time to take a look.

**Comment by huanghongxun** [comment]
@cbracken rebased on main.

**Comment by cbracken** [comment]
Added some comments to https://github.com/flutter/flutter/pull/74814 covering the framework side of this work. 

I think the embedder side is much less likely to be affected by the framework-side design work. I've pulled this patch down locally and am playing with it a bit to get a feel for it while I review. Thanks again for your patience on this.

**Comment by cbracken** [comment]
I'm going to close this for the time being, in order to get it off the review queue until we've worked out the right approach to consolidating the API in a way that works for all platforms. That's a problem being actively hacked out right now on macOS and Windows/Linux are on our agenda next. I do think that some of the consolidation work in the framework PR is something we want to followup on though.

**Comment by CaptainDario** [comment]
@cbracken Sorry for pinging you, but I have been following the platform view implementations already for quite some time.
As MacOS now landed (awesome work!) and the webview is also available in the latest webview package, I wanted to ask if now this PR could be looked at again.

**Comment by robert-ancell** [comment]
I think this would have to be a new PR at this point as a lot of things have changed. There's also a lot of change occurring at the moment adding support for multi-view. However, getting platform views would be great too!

**Comment by wanjm** [comment]
@robert-ancell  where can i get the progress of multi-window of linux;

**Comment by robert-ancell** [comment]
> @robert-ancell where can i get the progress of multi-window of linux;

See https://github.com/flutter/flutter/issues/138178

**Pull request closed without merging**